specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,220 @@
+
+Block producers in the particular epoch exchange many kinds of messages. The two kinds that are relevant to the consensus are **Blocks** and **Approvals**. The approval contains five fields:
+
+```rust
+enum ApprovalInner {
+    Endorsement(BlockHash),
+    Skip(BlockHeight),
+}
+
+struct Approval {
+    inner: ApprovalInner,
+    target_height: BlockHeight,
+    signature: Signature,
+    account_id: AccountId
+}
+```
+
+Where the parameter of the `Endorsement` is the hash of the approved block, the parameter of the `Skip` is the height of the approved block, `target_height` is the specific height at which the approval can be used (an approval with a particular `target_height` can be only included in the `approvals` of a block that has `height = target_height`), `account_id` is the account of the block producer who created the approval, and `signature` is their signature on the tuple `(inner, target_height)`.
+
+## Approvals Requirements
+
+A block `B` at height `h` that has some other block `B'` as its previous block must logically contain approvals of a form described in the next paragraph from block producers whose cumulative stake exceeds 2/3 of the total stake in the current epoch, and in specific conditions described in section [epoch switches](#epoch-switches) also the approvals of the same form from block producers whose cumulative stake exceeds 2/3 of the total stake in the next epoch.
+
+If a block to be produced will have height `h` and previous block `B'`, the approvals locically included in it must be an `Endorsement` with the hash of `B'` if and only if `h == B'.height + 1`, otherwise it must be a `Skip` with the height of `B'`. See [this section](#approval-condition) below for details on why the endorsements must contain the hash of the previous block, and skips must contain the height.
+
+Note that since each approval that is logically stored in the block is the same for each block producer (except for the `account_id` of the sender and the `signature`), it is redundant to store the full approvals. Instead physically we only store the signatures of the approvals. The specific way they are stored is the following: we first fetch the ordered set of block producers from the current epoch. If the block is on the epoch boundary and also needs to include approvals from the next epoch (see [epoch switches](#epoch-switches)), we add new accounts from the new epoch
+
+```python
+def get_accounts_for_block_ordered(h, prev_block):
+    cur_epoch = get_next_block_epoch(prev_block)
+    next_epoch = get_next_block_next_epoch(prev_block)
+
+    account_ids = get_epoch_block_producers_ordered(cur_epoch)
+    if next_block_needs_approvals_from_next_epoch(prev_block):
+        for account_id in get_epoch_block_producers_ordered(next_epoch):
+            if account_id not in account_ids:
+                account_ids.append(account_id)
+
+    return account_ids
+```
```

Comment:
So, we take a sorted set of approvals from the current epoch, and then append the sorted set of approvals from the next epoch excluding those that are also present in the current epoch. Why not instead sort the union of both sets?

Reply:
Because the set that comes from the epoch manager is not "sorted" per se, it is in the ordrer of producing blocks. Not sorting here allows me not to leak any sorting key from the epoch manager into the consensus.